### PR TITLE
[FIX] auth_totp: allow support of mobile device for 2FA

### PR DESCRIPTION
**Pull request by res-odoo**
opw-2369433
opw-2354969

### CODE REVIEW

File: addons/auth_totp/tests/test_totp.py [hunk 1]
```diff
@@ -81,3 +83,61 @@ def test_totp_administration(self):
...
         self.start_tour('/', 'totp_login_disabled', login=None)
+
+    def _build_payload(self, params={}):
+        """
+        Helper to properly build jsonrpc payload
+        """
+        return {
+            'jsonrpc': '2.0',
+            'method': 'call',
+            'id': str(uuid4()),
```
Comment (xmo-odoo) [inline addons/auth_totp/tests/test_totp.py:94]: useless complexity, you can just pass in the number `0`.

File: addons/auth_totp/tests/test_totp.py [hunk 2]
```diff
@@ -81,3 +83,61 @@ def test_totp_administration(self):
         self.start_tour('/web', 'totp_tour_setup', login='demo')
         self.start_tour('/web', 'totp_admin_disables', login='admin')
         self.start_tour('/', 'totp_login_disabled', login=None)
+
+    def _build_payload(self, params={}):
```
Comment (xmo-odoo) [inline addons/auth_totp/tests/test_totp.py:87]: Why id there a default value? In which situation would that be useful? Also / alternatively, why not use kwargs?

File: addons/auth_totp/tests/test_totp.py [hunk 3]
```diff
@@ -81,3 +83,61 @@ def test_totp_administration(self):
...
+        self.url_open('/web/session/logout')
+
+        # 2. query login
+        payload = self._build_payload({
+            'db': get_db_name(),
+            'login': 'demo',
+            'password': 'demo',
+            'context': {},
+        })
+        response = self.url_open('/web/session/authenticate', data=json.dumps(payload), headers=headers)
```
Comment (xmo-odoo) [inline addons/auth_totp/tests/test_totp.py:113]: `json` kwarg does the conversion on its own and automatically sets the request's content-type.

File: addons/auth_totp/tests/test_totp.py [hunk 4]
```diff
@@ -81,3 +83,61 @@ def test_totp_administration(self):
...
+        self.assertEqual(response.status_code, 200)
```
Comment (xmo-odoo) [inline addons/auth_totp/tests/test_totp.py:114]: Not super useful as JSON-RPC endpoints never return anything other than 200 unless they're missing.

Also `raise_for_status()` is the usual way to *assert* success (sadly it's not chainable).

File: addons/auth_totp/tests/test_totp.py [hunk 5]
```diff
@@ -81,3 +83,61 @@ def test_totp_administration(self):
...
+        self.assertEqual(data['result']['uid'], None)
+
+        # 3. query totp_token empty
+        payload = self._build_payload({
+            'totp_token': ''
+        })
+        response = self.url_open('/web/session/authenticate/totp', data=json.dumps(payload), headers=headers)
+        self.assertEqual(response.status_code, 200)
+        data = response.json()
+        self.assertEqual(data['result']['error'], _("Invalid authentication code format."))
```
Comment (xmo-odoo) [inline addons/auth_totp/tests/test_totp.py:125]: per [jsonrpc spec](https://www.jsonrpc.org/specification#response_object), the error should logically be signaled by `data['error']` being set, and `data['result']` being unset: `data['result']` signals success.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
Imma need way more explanations there, and an actual commit message. I've no idea how the mobile client works, what its flow is, and why it needs a new endpoint.

**Comment by xmo-odoo** [comment]
@res-odoo do we close this as you implement the client-only version?

**Comment by adr-odoo** [comment]
Client-only version: https://github.com/odoo/mobile/pull/190

**Pull request closed without merging**